plugin/src/main/java/org/openrewrite/gradle/isolated/AndroidParser.kt

```diff
@@ -0,0 +1,178 @@
+package org.openrewrite.gradle.isolated
+
+import com.android.build.gradle.AppExtension
+import com.android.build.gradle.BaseExtension
+import com.android.build.gradle.LibraryExtension
+import com.android.build.gradle.TestExtension
+import com.android.build.gradle.api.BaseVariant
+import com.android.build.gradle.internal.api.TestedVariant
+import org.gradle.api.DomainObjectSet
+import org.gradle.api.Project
+import org.gradle.api.logging.Logger
+import org.openrewrite.ExecutionContext
+import org.openrewrite.SourceFile
+import org.openrewrite.gradle.RewriteExtension
+import org.openrewrite.java.internal.JavaTypeCache
+import org.openrewrite.kotlin.KotlinParser
+import org.openrewrite.style.NamedStyles
+import java.io.File
+import java.io.IOException
+import java.io.UncheckedIOException
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.PathMatcher
+import java.nio.file.Paths
+import java.util.stream.Collectors
+import java.util.stream.Stream
+
+class AndroidParser(
+    val logger: Logger,
+    val rewriteExtension: RewriteExtension,
+    val styles: List<NamedStyles>,
+    val baseDir: Path,
+) {
+
+    private val BaseExtension.variants: DomainObjectSet<out BaseVariant>?
+        get() = when (this) {
+            is AppExtension -> applicationVariants
+            is LibraryExtension -> libraryVariants
+            is TestExtension -> applicationVariants
+            else -> null
+        }
+
+    private val BaseVariant.testVariants: List<BaseVariant>
+        get() = if (this is TestedVariant) {
+            listOfNotNull(testVariant, unitTestVariant)
+        } else {
+            emptyList()
+        }
+
+    private fun getKotlinParser(javaTypeCache: JavaTypeCache): KotlinParser {
+        return KotlinParser.builder()
+            .styles(styles)
+            .typeCache(javaTypeCache)
+            .logCompilationWarningsAndErrors(rewriteExtension.getLogCompilationWarningsAndErrors())
+            .build()
+    }
+
+    fun isAndroidProject(subproject: Project): Boolean {
+        return subproject.extensions.findByType(BaseExtension::class.java) != null
+    }
+
+    fun parseAndroidSources(
+        subproject: Project,
+        exclusions: Collection<PathMatcher>,
+        alreadyParsed: MutableSet<Path>,
+        ctx: ExecutionContext
+    ): Pair<Stream<SourceFile>, Int> {
+
+        val variants = subproject.extensions.findByType(BaseExtension::class.java)?.variants ?: return Stream.empty<SourceFile>() to 0
+
+        val sources = variants.map { variant ->
```

Comment:
Each gradle module can have multiple variants, and each variant can have multiple sourcesets. `variant.sourceSets` is not a list of `org.gradle.api.tasks.SourceSet`'s, so the existing parsing code can't be shared. It is a list of [SourceProvider's](https://android.googlesource.com/platform/tools/build/+/ef84df96b7b1a26a96a37506d95e3d427ee830ed/builder-model/src/main/java/com/android/builder/model/SourceProvider.java)